src/charm.py

```diff
@@ -5,49 +5,427 @@
+        return smtp_credentials
+
+    def _get_s3_credentials(self) -> dict[str, str] | None:
+        """Get penpot s3 credentials from the s3 integration.
+
+        Returns:
+            Penpot s3 credentials.
+        """
+        relation = self.model.get_relation("s3")
+        if not relation or not relation.app:
+            return None
+        s3_data = self.s3.get_s3_connection_info()
+        if not s3_data or "access-key" not in s3_data:
+            return None
+        return {
+            "AWS_ACCESS_KEY_ID": s3_data["access-key"],
+            "AWS_SECRET_ACCESS_KEY": s3_data["secret-key"],
+            "PENPOT_ASSETS_STORAGE_BACKEND": "assets-s3",
+            "PENPOT_STORAGE_ASSETS_S3_REGION": s3_data.get("region", "us-east-1"),
+            "PENPOT_STORAGE_ASSETS_S3_BUCKET": s3_data["bucket"],
+            "PENPOT_STORAGE_ASSETS_S3_ENDPOINT": s3_data["endpoint"],
+        }
+
+    def _get_public_uri(self) -> str | None:
+        """Get penpot public URI.
+
+        Returns:
+            Penpot public URI.
+        """
+        return self.ingress.url
+
+    def _get_penpot_frontend_options(self) -> list[str]:
+        """Retrieve the penpot options for the penpot frontend.
+
+        Returns:
+            Penpot frontend options.
+        """
+        return sorted(
+            [
+                "enable-login-with-password",
+                "disable-registration",
+                "disable-onboarding-questions",
+            ]
+        )
+
+    def _get_penpot_backend_options(self) -> list[str]:
+        """Retrieve the penpot options for the penpot backend.
+
+        Returns:
+            Penpot backend options.
+        """
+        return sorted(
+            [
+                "enable-login-with-password",
+                "enable-prepl-server",
+                "disable-registration",
+                "disable-telemetry",
+                "disable-onboarding-questions",
+                "disable-log-emails",
+                ("enable" if self._get_smtp_credentials() else "disable") + "-smtp",
+            ]
+        )
+
+    def _get_local_resolver(self) -> str:
+        """Retrieve the current nameserver address being used.
+
+        Returns:
+            The address of the nameserver.
+        """
+        kube_dns = f"kube-dns.kube-system.svc.{self._get_kubernetes_cluster_domain()}"
+        try:
+            dns.resolver.resolve(kube_dns, search=True)
+            return kube_dns
+        except dns.exception.DNSException:
+            # resolvers like dns-over-https, not likely to happen in Kubernetes
+            return typing.cast(str, dns.resolver.Resolver().nameservers[0])
+
+    def _get_penpot_exporter_unit(self) -> str:
+        """Retrieve the name of the unit designated to run the penpot exporter.
+
+        Returns:
+            Exporter unit name.
+        """
+        relation = typing.cast(ops.Relation, self.model.get_relation("penpot_peer"))
```

Comment:
Out of curiosity: All units will have an exporter running but only one (the lowest unit number) will be elected as the exporter, is that it?

Reply:
Yes, this is similar to the Indico case; all units will have an exporter installed, but there can be only one exporter instance running at the same time. I chose a simpler approach than Indico's, which is to use the unit with the lowest unit number to run the exporter instance. But, I have left room to change the selection algorithm in the future.

Reply:
Nice, very creative. :clap: 
In that case, wouldn't be needed to observe relation-departed? If for some reason the lowest unit number departs, a new exporter should be assigned, right?